### ISSUE TITLE
Is there any limit to the number of bad PINs anyone can try?

### ISSUE THREAD

**Issue by pauser0000001** [jim17/memtype#7]
Is the usb device protected against too many bad pins? how? I have not seen it in the documentation.

Incredible project, by the way.

**Comment by oyzzo** [comment]
Hi @pauser0000001, 
there's no limit for the moment, such a protection is indeed basic to be secure against brute force attacks.
A possible solution could be 3 bad pins resulting into full memory erase. What do you think?

**Comment by pauser0000001** [comment]
As the pin guess is something that is not possible to do offline (unless the usb has been stolen), I think a less dramatic protection (to avoid to have to write the passwords in another media) could be also effective (if it is possible).
Only if possible, several pins could be needed in case of failure (configured at pin time creation).
After 3 bad pins, another pin is required (introduced by the owner at the time of pins creation). 
The official library should advise the usb has been attacked and two or more pins are needed.
If the second pin is again now introduced 3 times, and the user configured another pin, then now 3 pins are needed (probable 3 pins are enough for everybody), otherwise do a full memory erase.
What do you think?

**Comment by oyzzo** [comment]
ok, I've been thinking about it and I think I've got it clear now.

Flash size limits us a lot on the firmware side, so every byte must be taken into account... that's why we implemented the NOEKEON in assembly for example. We need to keep as much space as we can free for password storage, and our target space was 2K, we're almost there :( 

Protecting against PIN bruteforce attacks is a needed feature for sure. Now there're two options, the simple approach of erasing on 3 fails, or the more sophisticated one that you suggest.

From the flash size point of view, simple option wins.
From the security point of view, simple option wins too.

Now let's check if it could be a problem for some use case.

The 2 main use cases I can see are 1) a non professional use for personal credentials, and 2) a professional use for work credentials on cloud/servers/devices/webs/softwares etc...

In case you loose the credentials in use case 1) All software/webs/apps I can think of have the proper "forgotten password" option to easily solve it, so no big drama.

In use case 2) the export option of the GUI is highly recommended, to have a backup (encrypted!) of all credentials for all users. Relying on having the credentials in a single device is not a wise decision on a profesional environment.
